jakarta.jdt/org.eclipse.lsp4jakarta.jdt.core/src/main/java/org/eclipse/lsp4jakarta/jdt/core/websocket/WebSocketDiagnosticsCollector.java

```diff
@@ -228,32 +228,36 @@ private void onMessageWSMessageFormats(IType type, List<Diagnostic> diagnostics,
                             if (resolvedTypeName != null
                                     && !resolvedTypeName.equals(WebSocketConstants.SESSION_CLASS)) {
                                 WebSocketConstants.MESSAGE_FORMAT messageFormat = getMessageFormat(resolvedTypeName);
-                                boolean duplicateFound = false;
+                                IAnnotation duplicateFound = null;
                                 switch (messageFormat) {
                                 case TEXT:
-                                    if (onMessageTextUsed) {
-                                        duplicateFound = true;
+                                    if (onMessageTextUsed != null) {
+                                        duplicateFound = onMessageTextUsed;
                                     }
-                                    onMessageTextUsed = true;
+                                    onMessageTextUsed = annotation;
                                     break;
                                 case BINARY:
-                                    if (onMessageBinaryUsed) {
-                                        duplicateFound = true;
+                                    if (onMessageBinaryUsed != null) {
+                                        duplicateFound = onMessageBinaryUsed;
                                     }
-                                    onMessageBinaryUsed = true;
+                                    onMessageBinaryUsed = annotation;
                                     break;
                                 case PONG:
-                                    if (onMessagePongUsed) {
-                                        duplicateFound = true;
+                                    if (onMessagePongUsed != null) {
+                                        duplicateFound = onMessagePongUsed;
                                     }
-                                    onMessagePongUsed = true;
+                                    onMessagePongUsed = annotation;
                                     break;
                                 }
-                                if (duplicateFound) {
-                                    Diagnostic diagnostic = createDiagnostic(annotation, unit,
+                                if (duplicateFound != null) {
```

Comment:
It seems that this only handles when there are only 2 onMessage annotations, should we also take care when there are more than 2?

Reply:
I could try to generalize it to more than 2 although I don't think it's necessary. Pong message formats only take one class. Text and binary message formats allow for two different classes: String or Reader, and ByteBuffer or InputStream respectively.

This means that if for say we have 3 duplicate @OnMessage text methods, there will be one signature with a string, another with a Reader and the last one with either a String or Reader. This will most likely cause a regular java duplicate method error so they'll get another diagnostic. In the scenario that they changed the signature in another way, that third diagnostic might not appear but as soon as they fix one of the first duplicates, they'd get the message for the second.

Reply:
Makes sense!